plugins/clipboard/plugin.js

```diff
@@ -2353,9 +2377,26 @@
 				this.id = value;
 			}
 
-			try {
-				this.$.setData( type, value );
-			} catch ( e ) {}
+			if ( this.fallbackDataTransfer.isRequired() ) {
+				this.fallbackDataTransfer.setData( type, value );
+
+			} else {
+				try {
+					this.$.setData( type, value );
+				} catch ( e ) {}
+			}
+		},
+
+		/**
+		 * Sets dataTransfer id.
+		 *
+		 * @since 4.8
+		 * @param {String} id
+		 */
+		setId: function( id ) {
```

Comment:
I think that `setId` method should cover both the case when browser supports `clipboardIdDataType` as other types, and when it uses `Text` to keep the id (old IE).

Now we have id setting logic distributed in two places, which is counter intuitive. The second place is this leftover: https://github.com/ckeditor/ckeditor-dev/blob/10b8bea5fa99bdaa33b719ff43830428f6256815/plugins/clipboard/plugin.js#L2178-L2192

Although 

```js
// Check if ID is already created.
this.id = this.getData( clipboardIdDataType );
```

part should remain in the constructor, but nothing else.

Reply:
I would be in favor of making this property a private one, and use `{@link #id}` in the description to link the property.

Reply:
The problem here is that `setId` method works now more like `storeId`. And we would like to store id only if dataTransfer was created by certain events (copy, cut, dragstart). But all dataTransfer objects no matter what was initiating event should have id generated. And the code mention above is one generating and setting id, which is a little different than storing it. I see two solutions:
* generate id only for dataTransfer from events which should store it (what is the point for having id if it is not stored anywhere so would by retained later)
* leave logic as is, and rename `setId` to sth like `storeId`

I see the id is used in a few places outside dataTransfer:
* https://github.com/ckeditor/ckeditor-dev/blob/10b8bea5fa99bdaa33b719ff43830428f6256815/plugins/clipboard/plugin.js#L159
* https://github.com/ckeditor/ckeditor-dev/blob/10b8bea5fa99bdaa33b719ff43830428f6256815/plugins/clipboard/plugin.js#L2041
* https://github.com/ckeditor/ckeditor-dev/blob/10b8bea5fa99bdaa33b719ff43830428f6256815/plugins/clipboard/plugin.js#L2106

so it should be probably generated for every dataTransfer object created. WDYT @mlewand ?

Btw. If we are going to make it private we should introduce a getter because accessing it from outside (like `dataTransfer._id`) doesn't look good.

Reply:
> The problem here is (...)

Calling any `setData()` on dataTransfer doesn't make sense in events other than `copy`, `cut`, `dragstart`.

What we eventually could do is to attempt to get an id in the datatransfer constructor, and if not found then assign some (new one), but without storing it.

I think that you're right about the fact that method name is misleading, `storeId` sounds better.


As for renaming, we can't rename `dataTransfer.id` to `dataTransfer._id` as we'll break stuff, and that's not something we want to do.